### PR TITLE
chore: migrate expressionrouter tests to kongintegration

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Migrates `expressionrouter` tests to `kongintegration`.

```shell
 go test -run=TestExpressionsRouterMatchers_GenerateValidExpressions ./test/kongintegration -v
=== RUN   TestExpressionsRouterMatchers_GenerateValidExpressions
=== PAUSE TestExpressionsRouterMatchers_GenerateValidExpressions
=== CONT  TestExpressionsRouterMatchers_GenerateValidExpressions

**********************************************************************************************
Ryuk has been disabled for the current execution. This can cause unexpected behavior in your environment.
More on this: https://golang.testcontainers.org/features/garbage_collector/
**********************************************************************************************
2023/10/11 00:03:27 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 23.0.6
  API Version: 1.42
  Operating System: Alpine Linux v3.18
  Total Memory: 9940 MB
  Resolved Docker Host: unix:///Users/grzegorz.burzynski@konghq.com/.colima/docker.sock
  Resolved Docker Socket Path: /Users/grzegorz.burzynski@konghq.com/.colima/docker.sock
  Test SessionID: ea4e6e9b6dfd080d31c9feab4c8e98ec65401651d40a9cf37741ac296d2bd97f
  Test ProcessID: 0899ec18-4abb-4331-ab8e-87f16fa69862
2023/10/11 00:03:27 🐳 Creating container for image kong:latest
2023/10/11 00:03:27 ✅ Container created: d2b96d663342
2023/10/11 00:03:27 🐳 Starting container: d2b96d663342
2023/10/11 00:03:27 ✅ Container started: d2b96d663342
2023/10/11 00:03:27 🚧 Waiting for container id d2b96d663342 image: kong:latest. Waiting for: &{timeout:<nil> deadline:<nil> Strategies:[0x140000a9640 0x140000a9660]}
2023/10/11 00:03:29 🐳 Creating container for image kong/httpbin:0.1.0
2023/10/11 00:03:29 ✅ Container created: fa4549a12c74
2023/10/11 00:03:29 🐳 Starting container: fa4549a12c74
2023/10/11 00:03:29 ✅ Container started: fa4549a12c74
2023/10/11 00:03:29 🚧 Waiting for container id fa4549a12c74 image: kong/httpbin:0.1.0. Waiting for: &{Port:80/ timeout:<nil> PollInterval:100ms}
=== RUN   TestExpressionsRouterMatchers_GenerateValidExpressions/exact_match_on_path
=== RUN   TestExpressionsRouterMatchers_GenerateValidExpressions/exact_match_on_path_and_host
=== RUN   TestExpressionsRouterMatchers_GenerateValidExpressions/exact_match_on_path_and_wildcard_match_on_host
2023/10/11 00:03:46 🐳 Terminating container: fa4549a12c74
2023/10/11 00:03:46 🚫 Container terminated: fa4549a12c74
2023/10/11 00:03:46 🐳 Terminating container: d2b96d663342
2023/10/11 00:03:46 🚫 Container terminated: d2b96d663342
--- PASS: TestExpressionsRouterMatchers_GenerateValidExpressions (19.20s)
    --- PASS: TestExpressionsRouterMatchers_GenerateValidExpressions/exact_match_on_path (5.06s)
    --- PASS: TestExpressionsRouterMatchers_GenerateValidExpressions/exact_match_on_path_and_host (5.03s)
    --- PASS: TestExpressionsRouterMatchers_GenerateValidExpressions/exact_match_on_path_and_wildcard_match_on_host (5.04s)
PASS
ok      github.com/kong/kubernetes-ingress-controller/v2/test/kongintegration   19.759s

```

### CODE REVIEW

File: test/kongintegration/containers/httpbin.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/kong/kubernetes-ingress-controller/v2/test"
+)
+
+// HTTPBin represents a docker container running the `kong/httpbin`.
+type HTTPBin struct {
+	container testcontainers.Container
+}
+
+// NewHTTPBin spawns a docker container running the `kong/httpbin` which can be used for proxy routing testing.
+func NewHTTPBin(ctx context.Context, t *testing.T) HTTPBin {
```
Comment (programmer04) [inline test/kongintegration/containers/httpbin.go:23]: nit: I rather prefer moving it above `type HTTPBin` to have method `IP(..)` closer to struct

File: test/kongintegration/expression_router_test.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		name            string
+		matcher         atc.Matcher
+		matchRequests   []*http.Request
+		unmatchRequests []*http.Request
+	}{
+		{
+			name:    "exact match on path",
+			matcher: atc.NewPredicateHTTPPath(atc.OpEqual, "/foo"),
+			matchRequests: []*http.Request{
+				helpers.MustHTTPRequest(t, "GET", helpers.MustParseURL(t, "http://a.foo.com").String(), "foo", nil),
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:44]:
```suggestion
				helpers.MustHTTPRequest(t, http.MethodGet, helpers.MustParseURL(t, "http://a.foo.com").String(), "foo", nil),
```
and the same for other methods

File: test/kongintegration/expression_router_test.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			atc.ApplyExpression(r, tc.matcher, 1)
+			req, err := kongClient.NewRequest("POST", "/config", nil, marshalKongConfig(t, *s, *r))
+			require.NoError(t, err)
+
+			resp, err := kongClient.DoRAW(ctx, req)
+			require.NoError(t, err)
+			require.NoError(t, resp.Body.Close())
+			require.Equal(t, http.StatusCreated, resp.StatusCode)
+
+			// matched requests should access upstream service
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:109]:
```suggestion
			// Matched requests should access the upstream service.
```

File: test/kongintegration/expression_router_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+					}
+					resp.Body.Close()
+					if resp.StatusCode != http.StatusOK {
+						return false
+					}
+				}
+				return true
+			}, time.Minute, 5*time.Second)
+
+			// unmatched requests should get a 404 from Kong
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:125]:
```suggestion
			// Unmatched requests should get a 404 from Kong.
```

File: test/kongintegration/expression_router_test.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+				require.Equal(t, http.StatusNotFound, resp.StatusCode)
+			}
+		})
+	}
+}
+
+func marshalKongConfig(t *testing.T, s kong.Service, r kong.Route) io.Reader {
+	t.Helper()
+	content := &file.Content{
+		FormatVersion: "3.0",
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:139]:
```suggestion
		FormatVersion: versions.DeckFileFormatVersion,
```

File: test/kongintegration/expression_router_test.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+			unmatchRequests: []*http.Request{
+				helpers.MustHTTPRequest(t, "GET", helpers.MustParseURL(t, "http://a.foo.com").String(), "foobar", nil),
+				helpers.MustHTTPRequest(t, "GET", helpers.MustParseURL(t, "http://a.bar.com").String(), "foo", nil),
+			},
+		},
+	}
+
+	proxyParsedURL, err := url.Parse(kongC.ProxyURL(ctx, t))
+	require.NoError(t, err)
+	proxyClient := helpers.DefaultHTTPClient()
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:84]: Can be replaced with `helpers.DefaultHTTPClientWithProxy(..)`?

File: test/kongintegration/containers/kong.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+// Kong represents a docker container running Kong.
+type Kong struct {
+	container testcontainers.Container
+}
+
+// NewKong spawns a docker container running Kong (its image is determined by environment variables).
+// It sets up a cleanup function that will terminate the container when the test finishes.
+func NewKong(ctx context.Context, t *testing.T) Kong {
```
Comment (programmer04) [inline test/kongintegration/containers/kong.go:29]: nit: I rather prefer moving it above type `Kong` to have method IP(..) closer to the struct

File: test/kongintegration/containers/httpbin.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	httpBinC, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		err := httpBinC.Terminate(ctx)
+		assert.NoError(t, err)
```
Comment (programmer04) [inline test/kongintegration/containers/httpbin.go:39]:
```suggestion
		assert.NoError(t, httpBinC.Terminate(ctx))
```

File: test/kongintegration/containers/kong.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	kongC, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		err := kongC.Terminate(ctx)
+		assert.NoError(t, err)
```
Comment (programmer04) [inline test/kongintegration/containers/kong.go:52]:
```suggestion
		assert.NoError(t, kongC.Terminate(ctx))
```

File: test/kongintegration/expression_router_test.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/kong/kubernetes-ingress-controller/v2/test/kongintegration/containers"
+)
+
+func TestExpressionsRouterMatchers_GenerateValidExpressions(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	kongC := containers.NewKong(ctx, t)
+	kongClient, err := kong.NewClient(lo.ToPtr(kongC.AdminURL(ctx, t)), &http.Client{})
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:29]:
```suggestion
	kongClient, err := kong.NewClient(lo.ToPtr(kongC.AdminURL(ctx, t)), helpers.DefaultHTTPClient())
```

File: test/kongintegration/expression_router_test.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+			req, err := kongClient.NewRequest("POST", "/config", nil, marshalKongConfig(t, *s, *r))
+			require.NoError(t, err)
+
+			resp, err := kongClient.DoRAW(ctx, req)
+			require.NoError(t, err)
+			require.NoError(t, resp.Body.Close())
+			require.Equal(t, http.StatusCreated, resp.StatusCode)
+
+			// matched requests should access upstream service
+			require.Eventually(t, func() bool {
```
Comment (programmer04) [inline test/kongintegration/expression_router_test.go:110]: We have a helper `EventuallyGETPath(...)`, maybe it can be used here. WDYT?

**Review by programmer04 — CHANGES_REQUESTED** [review]
Great effort 🎖️ just some nits

File: test/kongintegration/expression_router_test.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+						t.Logf("error happened on getting response from kong: %v", err)
+						return false
+					}
+					resp.Body.Close()
+					if resp.StatusCode != http.StatusOK {
+						return false
+					}
+				}
+				return true
+			}, time.Minute, 5*time.Second)
```
Comment (pmalek) [inline test/kongintegration/expression_router_test.go:123]: 5s is pretty generous, isn't it?

```suggestion
			}, time.Minute, time.Second)
```

File: Makefile [hunk 13]
```diff
@@ -474,14 +474,6 @@ test.istio: gotestsum
 		-timeout $(E2E_TEST_TIMEOUT) \
 		./test/e2e/...
 
-.PHONY: test.expression_router
-test.expression_router: gotestsum
```
Comment (pmalek) [inline Makefile:478]: I'm surprised that wasn't included anywhere in the CI.

**Review by pmalek — DISMISSED** [review]

File: test/kongintegration/containers/httpbin.go (hunk 1 above)
Comment (czeslavo) [reply test/kongintegration/containers/httpbin.go:23]: 🤔 What would be the benefit of this? I think that we rather always stick to the following order:

1. struct definition
2. its constructor(s)
3. other methods

Comment (pmalek) [reply test/kongintegration/containers/httpbin.go:23]: > 🤔 What would be the benefit of this? I think that we rather always stick to the following order:
> 
> 1. struct definition
> 2. its constructor(s)
> 3. other methods

Agree, I believe this is the best we can do with Go.

File: Makefile (hunk 13 above)
Comment (czeslavo) [reply Makefile:478]: Yeah, apparently we missed that. 😿

File: test/kongintegration/containers/httpbin.go (hunk 1 above)
Comment (programmer04) [reply test/kongintegration/containers/httpbin.go:23]: Yeah, I looked at the codebase and this is the most popular order in KIC. It's just a preference, even stdlib is not consistent about order. It's not something that I would argue about, let's do it as everywhere

**Review by pmalek — APPROVED** [review]
Very nice improvement 👍

**Review by programmer04 — APPROVED** [review]

**Pull request merged**